File: backend/models/domain.py
from datetime import datetime
from typing import Literal
import uuid

from pydantic import BaseModel, Field

from lib.dates import now_utc
from models.identity import PatientProfile, ProfileUpdate


ReferralStatus = Literal[
    "DRAFT",
    "CONFIRMED",
    "SENT",
    "RECEIVED",
    "ACCEPTED",
    "ARRIVED",
    "COMPLETED",
    "REJECTED",
    "REDIRECTED",
    "CANCELLED",
]


class Event(BaseModel):
    label: str
    timestamp: datetime = Field(default_factory=now_utc)
    detail: str | None = None
    actor_user_id: str | None = None
    actor_role: str | None = None


class AssessmentSessionCreate(BaseModel):
    symptoms: str = Field(min_length=3, max_length=2000)


class AssessmentMessage(BaseModel):
    session_id: str
    answer: str = Field(min_length=1, max_length=500)


class AssessmentSession(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    patient_id: str
    symptoms: str
    answers: list[str] = Field(default_factory=list)
    question_index: int = 0
    status: Literal["active", "completed"] = "active"
    outcome: str | None = None
    guidance: str | None = None
    created_at: datetime = Field(default_factory=now_utc)


class AssessmentMessageResponse(BaseModel):
    session_id: str
    question: str | None = None
    question_index: int
    total_questions: int
    complete: bool
    outcome: str | None = None
    guidance: str | None = None
    urgency: Literal["ROUTINE", "URGENT", "EMERGENCY"] = "ROUTINE"
    review: str = "deterministic safety-reviewed pathway"


class DoctorAssessmentRequest(BaseModel):
    access_request_id: str
    symptoms: str = Field(min_length=3, max_length=2000)
    findings: str = Field(default="", max_length=2000)
    vitals: str = Field(default="", max_length=1000)


class DoctorAssessment(BaseModel):
    mode: Literal["deterministic_fallback"] = "deterministic_fallback"
    urgency: str = "ROUTINE"
    care_requirement: str = "Primary care review"
    summary: str
    suggested_capabilities: list[str]
    review: str = "AI/rule-assisted support only — clinician confirmation required"


class Facility(BaseModel):
    id: str
    name: str
    location: str
    distance: str
    capabilities: list[str]
    match_reason: str
    emergency_ready: bool = False


class FacilityMatchResponse(BaseModel):
    requirement: str
    facilities: list[Facility]
    note: str = "Mandatory capabilities first; location and fixed demo order are tie-breakers."


class ReferralCreate(BaseModel):
    access_request_id: str
    facility_id: str
    reason: str = Field(min_length=3, max_length=500)
    care_requirement: str = "Primary care review"
    urgency: str = "ROUTINE"
    symptoms: str = Field(default="", max_length=2000)
    findings: str = Field(default="", max_length=2000)
    vitals: str = Field(default="", max_length=1000)
    ai_assessment: str = Field(default="", max_length=2000)
    doctor_decision: str = Field(min_length=3, max_length=1000)
    idempotency_key: str = Field(min_length=8, max_length=100)


class ReferralPatientSnapshot(BaseModel):
    age: int
    gender: str
    blood_group: str | None = None
    existing_conditions: list[str] = Field(default_factory=list)
    allergies: list[str] = Field(default_factory=list)
    medicines: list[str] = Field(default_factory=list)


class Referral(BaseModel):
    id: str = Field(default_factory=lambda: f"REF-{uuid.uuid4().hex[:10].upper()}")
    idempotency_key: str
    patient_id: str
    patient_code: str
    patient_name: str
    patient_summary: ReferralPatientSnapshot
    access_request_id: str | None = None
    referring_user_id: str | None = None
    referring_provider_name: str = ""
    facility_id: str
    facility_name: str
    reason: str
    care_requirement: str
    urgency: str
    symptoms: str = ""
    findings: str = ""
    vitals: str = ""
    ai_assessment: str = ""
    doctor_decision: str
    status: ReferralStatus = "DRAFT"
    emergency: bool = False
    created_at: datetime = Field(default_factory=now_utc)
    updated_at: datetime = Field(default_factory=now_utc)
    events: list[Event] = Field(default_factory=list)
    outcome: str | None = None


class ReferralActionRequest(BaseModel):
    status: Literal["ACCEPTED", "REJECTED", "REDIRECTED", "ARRIVED", "COMPLETED", "CANCELLED"]
    outcome: str | None = Field(default=None, max_length=2000)
    reason: str | None = Field(default=None, max_length=1000)


class Document(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    patient_id: str
    filename: str
    storage_name: str
    content_type: str
    size_bytes: int
    review_status: Literal["processing", "needs_verification", "unreadable", "verified", "rejected"] = "processing"
    extracted_fields: list[str] = Field(default_factory=list)
    original_available: bool = True
    extraction_mode: str = "deterministic_mock"
    review_actor_user_id: str | None = None
    reviewed_at: datetime | None = None
    created_at: datetime = Field(default_factory=now_utc)


class DocumentReviewRequest(BaseModel):
    access_request_id: str
    decision: Literal["verified", "rejected"]


class SOSCreate(BaseModel):
    note: str = Field(default="", max_length=500)
    idempotency_key: str = Field(min_length=8, max_length=100)


class EmergencyEvent(BaseModel):
    id: str = Field(default_factory=lambda: f"SOS-{uuid.uuid4().hex[:10].upper()}")
    idempotency_key: str
    patient_id: str
    patient_code: str
    patient_name: str
    status: str = "AMBULANCE_INITIATED"
    dispatch_status: str = "SIMULATED_DISPATCH"
    note: str = ""
    selected_facility_id: str | None = None
    selected_facility_name: str | None = None
    referral_id: str | None = None
    created_at: datetime = Field(default_factory=now_utc)
    updated_at: datetime = Field(default_factory=now_utc)
    audit_note: str = "Emergency action and access events are audit logged."


class SOSResponse(BaseModel):
    event: EmergencyEvent
    facilities: list[Facility]
    message: str


class SOSFacilitySelect(BaseModel):
    sos_id: str
    facility_id: str


class SOSContextUpdate(BaseModel):
    note: str = Field(min_length=1, max_length=500)


class DashboardState(BaseModel):
    profile: PatientProfile | None = None
    facilities: list[Facility]
    referrals: list[Referral]
    emergencies: list[EmergencyEvent]
    documents: list[Document]


class HospitalQueue(BaseModel):
    emergencies: list[EmergencyEvent]
    referrals: list[Referral]

File: frontend/src/types/domain.ts
export type Role = "patient" | "doctor" | "health_worker" | "hospital_doctor" | "hospital_admin";
export type ReviewStatus = "processing" | "needs_verification" | "unreadable" | "verified" | "rejected";
export type ReferralStatus = "DRAFT" | "CONFIRMED" | "SENT" | "RECEIVED" | "ACCEPTED" | "ARRIVED" | "COMPLETED" | "REJECTED" | "REDIRECTED" | "CANCELLED";

export interface SessionUser {
  id: string;
  email: string;
  display_name: string;
  role: Role;
  patient_code?: string | null;
  facility_id?: string | null;
}

export interface LoginRequest { email: string; password: string }

export interface EventItem {
  label: string;
  timestamp: string;
  detail?: string | null;
  actor_user_id?: string | null;
  actor_role?: string | null;
}

export interface Profile {
  patient_code: string;
  name: string;
  age: number;
  gender: string;
  mobile: string;
  location: string;
  address: string;
  emergency_contact: string;
  blood_group?: string | null;
  weight_kg?: number | null;
  existing_conditions: string[];
  allergies: string[];
  medicines: string[];
  previous_history: string[];
  profile_complete: boolean;
}

export interface ProfileUpdate {
  name: string;
  age: number;
  gender: string;
  mobile: string;
  location: string;
  address: string;
  emergency_contact: string;
  blood_group?: string | null;
  weight_kg?: number | null;
  existing_conditions: string[];
  allergies: string[];
  medicines: string[];
  previous_history: string[];
}

export interface Facility {
  id: string;
  name: string;
  location: string;
  distance: string;
  capabilities: string[];
  match_reason: string;
  emergency_ready: boolean;
}

export interface Referral {
  id: string;
  idempotency_key: string;
  patient_id: string;
  patient_code: string;
  patient_name: string;
  patient_summary: {
    age: number;
    gender: string;
    blood_group?: string | null;
    existing_conditions: string[];
    allergies: string[];
    medicines: string[];
  };
  access_request_id?: string | null;
  referring_user_id?: string | null;
  referring_provider_name: string;
  facility_id: string;
  facility_name: string;
  reason: string;
  care_requirement: string;
  urgency: string;
  symptoms: string;
  findings: string;
  vitals: string;
  ai_assessment: string;
  doctor_decision: string;
  status: ReferralStatus;
  emergency: boolean;
  created_at: string;
  updated_at: string;
  events: EventItem[];
  outcome?: string | null;
}

export interface ReferralCreate {
  access_request_id: string;
  facility_id: string;
  reason: string;
  care_requirement: string;
  urgency: string;
  symptoms: string;
  findings: string;
  vitals: string;
  ai_assessment: string;
  doctor_decision: string;
  idempotency_key: string;
}

export interface ReferralActionRequest {
  status: "ACCEPTED" | "REJECTED" | "REDIRECTED" | "ARRIVED" | "COMPLETED" | "CANCELLED";
  outcome?: string;
  reason?: string;
}

export interface EmergencyEvent {
  id: string;
  idempotency_key: string;
  patient_id: string;
  patient_code: string;
  patient_name: string;
  status: string;
  dispatch_status: string;
  note: string;
  selected_facility_id?: string | null;
  selected_facility_name?: string | null;
  referral_id?: string | null;
  created_at: string;
  updated_at: string;
  audit_note: string;
}

export interface DocumentItem {
  id: string;
  patient_id: string;
  filename: string;
  storage_name: string;
  content_type: string;
  size_bytes: number;
  review_status: ReviewStatus;
  extracted_fields: string[];
  original_available: boolean;
  extraction_mode: string;
  review_actor_user_id?: string | null;
  reviewed_at?: string | null;
  created_at: string;
}

export interface DashboardState {
  profile?: Profile | null;
  facilities: Facility[];
  referrals: Referral[];
  emergencies: EmergencyEvent[];
  documents: DocumentItem[];
}

export interface AssessmentStart {
  session_id: string;
  question: string | null;
  question_index: number;
  total_questions: number;
  complete: boolean;
  urgency: "ROUTINE" | "URGENT" | "EMERGENCY";
}

export interface AssessmentAnswer extends AssessmentStart {
  outcome?: string | null;
  guidance?: string | null;
}

export interface DoctorAssessment {
  mode: "deterministic_fallback";
  urgency: string;
  care_requirement: string;
  summary: string;
  suggested_capabilities: string[];
  review: string;
}

export interface SOSResponse { event: EmergencyEvent; facilities: Facility[]; message: string }
export interface SOSCreate { note: string; idempotency_key: string }

export interface PatientIdentity {
  patient_code: string;
  normalized_phone: string;
  qr_payload: string;
  qr_svg_data_url: string;
  qr_expires_at: string;
}

export interface ConsentCodeResponse { consent_code: string; expires_at: string; instruction: string }
export interface IdentityLookupRequest { method: "patient_code" | "phone" | "qr"; value: string }

export interface IdentityPreview {
  access_request_id: string;
  patient_code: string;
  name: string;
  age: number;
  gender: string;
  masked_phone: string;
  method: string;
  expires_at: string;
  authorization_required: boolean;
}

export interface AuthorizedPatientAccess {
  access_request_id: string;
  patient: Profile;
  authorization_granted: boolean;
  expires_at: string;
  access_note: string;
  documents: DocumentItem[];
  prior_referrals: Referral[];
}

export const EMPTY_STATE: DashboardState = {
  profile: null,
  facilities: [],
  referrals: [],
  emergencies: [],
  documents: [],
};